src/app/calculator/furnaces/efficiency-improvement/efficiency-improvement-graph/efficiency-improvement-graph.component.ts

```diff
@@ -0,0 +1,331 @@
+import { ChangeDetectorRef, Component, ElementRef, HostListener, Input, OnInit, SimpleChanges, ViewChild } from '@angular/core';
+import { SimpleChart, TraceCoordinates } from '../../../../shared/models/plotting';
+
+import * as Plotly from 'plotly.js';
+import { Settings } from '../../../../shared/models/settings';
+import { graphColors } from '../../../../phast/phast-report/report-graphs/graphColors';
+import { EfficiencyImprovementInputs, EfficiencyImprovementOutputs } from '../../../../shared/models/phast/efficiencyImprovement';
+import { Axis, DisplayPoint, EfficiencyImprovementGraphService } from '../efficiency-improvement-graph.service';
+
+@Component({
+  selector: 'app-efficiency-improvement-graph',
+  templateUrl: './efficiency-improvement-graph.component.html',
+  styleUrls: ['./efficiency-improvement-graph.component.css']
+})
+export class EfficiencyImprovementGraphComponent implements OnInit {
+
+  @Input()
+  settings: Settings;
+  @Input()
+  efficiencyImprovementInputs: EfficiencyImprovementInputs;
+  @Input()
+  reset: boolean;
+  @Input()
+  efficiencyImprovementOutputs: EfficiencyImprovementOutputs;
+  
+  // DOM
+  @ViewChild("ngChartContainer", { static: false }) ngChartContainer: ElementRef;
+  @ViewChild('dataSummaryTable', { static: false }) dataSummaryTable: ElementRef;
+  dataSummaryTableString: any;
+  tabPanelChartId: string = 'tabPanelDiv';
+  expandedChartId: string = 'expandedChartDiv';
+  currentChartId: string = 'tabPanelDiv';
+
+  // Graph Data
+  selectedDataPoints: Array<DisplayPoint>;
+  defaultSelectedPointsData: Array<DisplayPoint>;
+  efficiencyChart: SimpleChart;
+  selectedAxisOptions = Array<{display: string, value: number}>();
+  selectedAxis: number = 0;
+  xAxis: Axis;
+
+  // Tooltips
+  hoverBtnGridLines: boolean = false;
+  displayGridLinesTooltip: boolean = false;
+  hoverBtnExpand: boolean = false;
+  displayExpandTooltip: boolean = false;
+  hoverBtnCollapse: boolean = false;
+  displayCollapseTooltip: boolean = false;
+  expanded: boolean = false;
+
+  @HostListener('document:keyup', ['$event'])
+  closeExpandedGraph(event) {
+    if (this.expanded) {
+      if (event.code === 'Escape') {
+        this.contractChart();
+      }
+    }
+  }
+
+  constructor(private cd: ChangeDetectorRef,
+              private efficiencyImprovementGraphService: EfficiencyImprovementGraphService) { }
+
+  ngOnInit() {
+    this.initAxisOptions();
+    this.triggerInitialResize();
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (!changes.efficiencyImprovementInputs.firstChange && !this.reset) {
```

Comment:
This can cause problems. `efficiencyImprovementInputs` may not end up part of the changes object. You can just add `changes.efficiencyImprovementInputs && ..`  prior to rest of if statement. It may not be an issue here but is something I have run into before.